### PR TITLE
Add async checkpointing for KD distributed and lora dpo distributed

### PR DESCRIPTION
**Pull request by ankitageorge**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Add async checkpointing for KD distributed and lora dpo distributed recipes
* Also fix how teacher checkpoints are initialized with checkpoint client so that it doesn't involve hacky-ness of updating the config manually

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x ] run unit tests via `pytest tests`
- [x ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/knowledge_distillation_single_device.py [hunk 1]
```diff
@@ -149,17 +149,11 @@ def load_teacher_checkpoint(self, cfg: DictConfig) -> dict[str, Any]:
...
-        teacher_checkpoint_val = new_cfg.pop("teacher_checkpointer")
-        new_cfg["checkpointer"] = teacher_checkpoint_val
-
-        teacher_checkpoint_client = CheckpointClient(
-            new_cfg,
+
+        teacher_checkpointer = config.instantiate(
+            cfg.teacher_checkpointer,
         )
+        teacher_checkpoint_client = CheckpointClient(cfg, teacher_checkpointer)
```
Comment (joecummings) [inline recipes/knowledge_distillation_single_device.py:156]: IIUC, because we initialize with a checkpointer, when we go to load in a checkpoint, it will hit [this](https://github.com/pytorch/torchtune/blob/3e353259091ccdde439dbcc150812859319c4d81/torchtune/training/checkpointing/_checkpoint_client.py#L93) line and NOT load in a distributed checkpoint with a recipe state.

Is that right?

**Review by joecummings — COMMENTED** [review]
Just a question

File: recipes/knowledge_distillation_single_device.py (hunk 1 above)
Comment (ankitageorge) [reply recipes/knowledge_distillation_single_device.py:156]: yes, you're right. But from my understanding we would never save an intermediate teacher checkpoint right? so it would never be saved with DCP anyways

**Review by joecummings — APPROVED** [review]
